### PR TITLE
Redesign h2policy

### PR DESCRIPTION
**Pull request by hazemakhalek**
## Changes proposed in this Pull Request
This PR redesings the H2 policy features already existing and adds a new dimension to it. In this PR, one can separate between the different H2 colors, Green/Yellow based on the temporal matching applied, Pink coming from nuclear plants and Blue from SMR with CC.

## Checklist

- [x] I tested my contribution locally and it seems to work fine.
- [x] Code and workflow changes are sufficiently documented.
- [x] Newly introduced dependencies are added to `envs/environment.yaml` and `envs/environment.docs.yaml`.

### CODE REVIEW

**Comment by hazemakhalek** [comment]
> @hazemakhalek thanks for pushing this forward! :) I also have a proposal for a redesign I have done a while ago here: https://github.com/pypsa-meets-earth/pypsa-earth-sec/tree/mar_paper, in particular [0431ec0](https://github.com/pypsa-meets-earth/pypsa-earth-sec/commit/0431ec0b9a515d3ebc8af5a9305415eb273a5ada). We should align these approaches if possible.

Hey Leon, I think we can easily adapt it to include your additions. Do you know already when will you be back?

**Comment by energyLS** [comment]
@hazemakhalek back now, happy to chat about it :)

File: scripts/solve_network.py [hunk 1]
```diff
@@ -517,17 +583,12 @@ def add_existing(n):
             add_existing(n)
 
         if (
-            snakemake.config["policy_config"]["monthly"]["reference_case"]
-            and eval(snakemake.wildcards["h2export"]) != 0
-            and snakemake.config["policy_config"]["policy"]
-            == "H2_export_monthly_constraint"
+            snakemake.config["policy_config"]["hydrogen"]["additionality"]
+            and snakemake.config["policy_config"]["hydrogen"]["temporal_matching"]
+            != "no_res_matching"
```
Comment (Eddy-JV) [inline scripts/solve_network.py:588]: Shouldn't also `is_reference` in config be false?

File: scripts/prepare_sector_network.py [hunk 2]
```diff
@@ -307,30 +383,65 @@ def add_hydrogen(n, costs):
...
+                # lifetime=costs.at["battery inverter", "lifetime"],
+            )
+
+            n.madd(
+                "Link",
+                nodes + " H2 UHS discharger",
+                bus0=nodes,
+                bus1=nodes + " H2 UHS",
+                carrier="H2 UHS discharger",
+                efficiency=0.83,
```
Comment (Eddy-JV) [inline scripts/prepare_sector_network.py:426]: Make it consistent between consum data and none custom data.

File: scripts/prepare_sector_network.py [hunk 3]
```diff
@@ -307,30 +383,65 @@ def add_hydrogen(n, costs):
...
+                # lifetime=costs.at["battery inverter", "lifetime"],
+            )
+
     # hydrogen stored overground (where not already underground)
     h2_capital_cost = costs.at[
         "hydrogen storage tank type 1 including compressor", "fixed"
     ]
     nodes_overground = cavern_nodes.index.symmetric_difference(nodes)
-
+    nodes_overground = nodes
```
[the remark this answers is not in the document]
Comment (Eddy-JV) [reply scripts/prepare_sector_network.py:437]: Remove line 436

[the remark this answers is not in the document]
Comment (hazemakhalek) [reply scripts/prepare_sector_network.py:437]: Yes, it's better to include the storage tanks everywhere just like any other technology.

File: scripts/prepare_sector_network.py [hunk 4]
```diff
@@ -270,27 +304,69 @@ def add_hydrogen(n, costs):
...
                 capital_cost=h2_capital_cost,
             )
 
+            n.madd(
+                "Link",
+                nodes + " H2 UHS charger",
+                bus0=nodes + " H2",
+                bus1=nodes + " H2 UHS",
+                carrier="H2 UHS charger",
+                # efficiency=costs.at["battery inverter", "efficiency"] ** 0.5,
```
[the remark this answers is not in the document]
Comment (hazemakhalek) [reply scripts/prepare_sector_network.py:352]: put everything to 1, yes that's a whole different issue that someone should look into

File: scripts/prepare_sector_network.py [hunk 5]
```diff
@@ -307,30 +383,65 @@ def add_hydrogen(n, costs):
...
+            )
+
             n.madd(
                 "Store",
-                cavern_nodes.index + " H2 Store",
-                bus=cavern_nodes.index + " H2",
+                cavern_nodes.index + " H2 UHS",
+                bus=cavern_nodes.index + " H2 UHS",
                 e_nom_extendable=True,
                 e_nom_max=h2_pot.values,
```
[the remark this answers is not in the document]
Comment (hazemakhalek) [reply scripts/prepare_sector_network.py:402]: Done

File: scripts/prepare_sector_network.py [hunk 6]
```diff
@@ -1096,7 +1369,18 @@ def add_shipping(n, costs):
             p_set=ports["p_set"],
         )
 
-        co2 = ports["p_set"].sum() * costs.at["oil", "CO2 intensity"]
+        if snakemake.config["sector"]["international_bunkers"]:
+            co2 = ports["p_set"].sum() * costs.at["oil", "CO2 intensity"]
+        else:
+            domestic_to_international = (
+                energy_totals["total international navigation"]
```
[the remark this answers is not in the document]
Comment (hazemakhalek) [reply scripts/prepare_sector_network.py:1376]: done

File: scripts/solve_network.py [hunk 7]
```diff
@@ -383,29 +378,100 @@ def add_co2_sequestration_limit(n, sns):
...
+    if (
+        snakemake.config["policy_config"]["hydrogen"]["temporal_matching"]
+        == "h2_yearly_matching"
+    ):
+        if snakemake.config["policy_config"]["hydrogen"]["additionality"] == True:
+            logger.info(
+                "additionality is currently not supported for yearly constraints, proceeding without additionality"
+            )
         logger.info("setting h2 export to yearly greenness constraint")
         H2_export_yearly_constraint(n)
```
[the remark this answers is not in the document]
Comment (hazemakhalek) [reply scripts/solve_network.py:442]: To be dealt with in a separate issue

File: scripts/prepare_sector_network.py (hunk 2 above)
Comment (hazemakhalek) [reply scripts/prepare_sector_network.py:426]: Dealt with

File: scripts/solve_network.py (hunk 1 above)
Comment (hazemakhalek) [reply scripts/solve_network.py:588]: That's right
Done

File: scripts/prepare_sector_network.py [hunk 8]
```diff
@@ -862,7 +1043,18 @@ def add_aviation(n, cost):
         p_set=airports["p_set"],
     )
 
-    co2 = airports["p_set"].sum() * costs.at["oil", "CO2 intensity"]
+    if snakemake.config["sector"]["international_bunkers"]:
+        co2 = airports["p_set"].sum() * costs.at["oil", "CO2 intensity"]
+    else:
+        domestic_to_international = (
+            energy_totals["total international aviation"]
```
[the remark this answers is not in the document]
Comment (hazemakhalek) [reply scripts/prepare_sector_network.py:1052]: Done

**Comment by Eddy-JV** [comment]
It's fine from my side. WE can merge. But I guess you have to resolve the merge issue now due to the merge of africa's PR yesterday. I am happy to do it together if necessary.

**Pull request merged**